### PR TITLE
do not create empty epochs

### PR DESCRIPTION
**Pull request by jkitman**
Changes consensus to stop generating empty epochs, instead only generate epochs when a transaction is received or a module has a non-empty proposal.
- At idle the federation consumes almost no CPU/bandwidth which could help running on lightweight servers (see #149)
- The consensus code is simplified because we do not have separate HBBFT threads (prepares us for #148)
- Will allow us to download/store consensus history without an enormous amount of empty epochs (see #142)
- The latency from when a user requests to mint coins / pay an LN invoice to when the federation completes the action is greatly improved, providing a better user experience:

```
Latency test 3 members
Reissue: 8.01s ---> 1.09s
Pay invoice: 8.15s ---> 3.03s

Latency test 6 members (max evil 1)
Reissue: 8.51s ---> 1.09s
Pay invoice: 7.58s ---> 3.02s

Latency test 9 members (max evil 2)
Reissue: 8.62s ---> 1.09s
Pay invoice: 6.84s ---> 3.02s
```

### CODE REVIEW

File: minimint/src/lib.rs [hunk 1]
```diff
@@ -41,240 +44,180 @@ pub mod config;
...
+            outcomes = self.handle_message(msg).await;
+        }
+        outcomes
+    }
+
+    async fn await_proposal_or_peer_message(&mut self) -> Option<PeerMessage> {
+        let awaits: Vec<Pin<Box<dyn Future<Output = Option<PeerMessage>>>>> = vec![
+            Box::pin(self.consensus.transaction_notify.notified().map(|_| None)),
+            Box::pin(self.consensus.await_consensus_proposal().map(|_| None)),
+            Box::pin(self.connections.receive().map(Some)),
```
Comment (elsirion) [inline minimint/src/lib.rs:198]: Can we do better than creating three heap allocations per received message here? I think I did the same in other places, just wondering if there might be an easy way to pin on the stack here.

File: modules/minimint-wallet/src/lib.rs [hunk 2]
```diff
@@ -1206,7 +1225,7 @@ pub enum WalletError {
     #[error("Unknown block hash in peg-in proof: {0}")]
     UnknownPegInProofBlock(BlockHash),
     #[error("Invalid peg-in proof: {0}")]
-    PegInProofError(#[from] PegInProofError),
+    PegInProofError(PegInProofError),
```
Comment (elsirion) [inline modules/minimint-wallet/src/lib.rs:1228]: Any particular reason for reversing this recent change or just a rebase artifact?

File: scripts/latencytest.sh [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+time for i in {1..10..1}
+do
+  echo "REISSUE $i"
+  TOKENS=$($MINT_CLIENT spend 1000)
+  $MINT_CLIENT reissue $TOKENS
+  $MINT_CLIENT fetch
+done
+
+## outgoing lightning
+time for i in {1..10..1}
```
Comment (elsirion) [inline scripts/latencytest.sh:107]: I don't understand what the `..1` does in the end here, I compared `for i in {1..10..1}` and `for i in {1..10}` in my local shell and it appears to be identical? Seems to be just the step size that defaults to 1 anyway.

**Review by elsirion — CHANGES_REQUESTED** [review]
Really great PR, just want to make sure that one `#[from]` change wasn't accidental.

I'm still wondering how you ended up with such bad benchmark numbers. I think in general you could divide them by 2 since users generally care more about transactions being final than receiving tokens, but that would still mean 4s in the original example. Anyway, this PR seems to drastically improve that :+1:

File: modules/minimint-wallet/src/lib.rs (hunk 2 above)
Comment (jkitman) [reply modules/minimint-wallet/src/lib.rs:1228]: I must have missed this...for some reason `cargo +nightly fix` wanted to remove these.  I'll add it back.

File: scripts/latencytest.sh (hunk 3 above)
Comment (jkitman) [reply scripts/latencytest.sh:107]: Yes, it's the step size defaults to 1...suppose I could remove.

File: modules/minimint-wallet/src/lib.rs [hunk 4]
```diff
@@ -165,13 +168,24 @@ impl FederationModule for Wallet {
...
     type VerificationCache = ();
 
+    async fn await_consensus_proposal<'a>(&'a self, rng: impl RngCore + CryptoRng + 'a) -> () {
+        let mut our_target_height = self.target_height().await;
+        let last_consensus_height = self.consensus_height().unwrap_or(0);
+
+        if self.consensus_proposal(rng).await.len() == 1 {
+            while our_target_height <= last_consensus_height {
+                our_target_height = self.target_height().await;
+                sleep(Duration::from_millis(100)).await;
```
[the remark this answers is not in the document]
Comment (jkitman) [reply modules/minimint-wallet/src/lib.rs:178]: 1 second should be sufficient I'll change that.

I'm not familiar with ZMQ, tried to use the bitcoin rpc but it seems to fail.

**Comment by jkitman** [comment]
> I'm still wondering how you ended up with such bad benchmark numbers. I think in general you could divide them by 2 since users generally care more about transactions being final than receiving tokens, but that would still mean 4s in the original example. Anyway, this PR seems to drastically improve that +1

I think the latency is due to the 2 sec epoch creation interval combined with generating empty epochs every other epoch.  With higher network latency the values might actually be worse, for instance if the federation is running over Tor.

For minted tokens you'd be done spending once the tx is accepted, but still need to remain online to fetch the signed tokens in the 2nd epoch.

The LN spend latency is more of the pain point if the user is trying to pay a merchant in person.  Oddly, the latency seemed to improve slightly with more federation members.

**Review by elsirion — APPROVED** [review]

**Pull request merged**